File: packages/server/src/clients/connection-manager.ts
import type { Logger } from "../logger"

const STALE_CONNECTION_TIMEOUT_MS = 45000
const STALE_SWEEP_INTERVAL_MS = 5000

export interface ClientConnectionRef {
  clientId: string
  connectionId: string
}

export interface ClientConnectionRecord extends ClientConnectionRef {
  key: string
  connectedAt: number
  lastSeenAt: number
}

type ConnectionChangeEvent = {
  type: "connected" | "disconnected"
  connection: ClientConnectionRecord
  reason?: string
}

interface RegisteredConnection extends ClientConnectionRecord {
  close: () => void
}

export class ClientConnectionManager {
  private readonly connections = new Map<string, RegisteredConnection>()
  private readonly subscribers = new Set<(event: ConnectionChangeEvent) => void>()
  private readonly sweepTimer: NodeJS.Timeout

  constructor(private readonly logger: Logger) {
    this.sweepTimer = setInterval(() => this.sweepStaleConnections(), STALE_SWEEP_INTERVAL_MS)
    this.sweepTimer.unref?.()
  }

  shutdown(): void {
    clearInterval(this.sweepTimer)
    for (const connection of Array.from(this.connections.values())) {
      this.disconnect(connection.key, "shutdown", false)
    }
  }

  subscribe(listener: (event: ConnectionChangeEvent) => void): () => void {
    this.subscribers.add(listener)
    return () => this.subscribers.delete(listener)
  }

  register(input: ClientConnectionRef & { close: () => void }): () => void {
    const key = getConnectionKey(input)
    const now = Date.now()
    const existing = this.connections.get(key)

    if (existing) {
      this.logger.debug({ clientId: input.clientId, connectionId: input.connectionId }, "Replacing existing client connection")
      this.disconnect(key, "replaced")
    }

    const connection: RegisteredConnection = {
      key,
      clientId: input.clientId,
      connectionId: input.connectionId,
      connectedAt: now,
      lastSeenAt: now,
      close: input.close,
    }
    this.connections.set(key, connection)
    this.logger.debug({ clientId: input.clientId, connectionId: input.connectionId }, "Client connected")
    this.notify({ type: "connected", connection })
    return () => this.disconnect(key, "closed")
  }

  pong(input: ClientConnectionRef): boolean {
    const key = getConnectionKey(input)
    const connection = this.connections.get(key)
    if (!connection) {
      this.logger.debug({ clientId: input.clientId, connectionId: input.connectionId }, "Ignoring pong for unknown client connection")
      return false
    }

    connection.lastSeenAt = Date.now()
    return true
  }

  isConnected(input: ClientConnectionRef): boolean {
    return this.connections.has(getConnectionKey(input))
  }

  private sweepStaleConnections(): void {
    const cutoff = Date.now() - STALE_CONNECTION_TIMEOUT_MS
    for (const connection of Array.from(this.connections.values())) {
      if (connection.lastSeenAt > cutoff) continue
      this.logger.debug({ clientId: connection.clientId, connectionId: connection.connectionId }, "Client connection timed out")
      this.disconnect(connection.key, "timeout")
    }
  }

  private disconnect(key: string, reason: string, invokeClose = true): void {
    const connection = this.connections.get(key)
    if (!connection) return
    this.connections.delete(key)
    this.logger.debug({ clientId: connection.clientId, connectionId: connection.connectionId, reason }, "Client disconnected")

    if (invokeClose) {
      try {
        connection.close()
      } catch (error) {
        this.logger.warn({ err: error, clientId: connection.clientId, connectionId: connection.connectionId }, "Failed to close stale client connection")
      }
    }

    this.notify({ type: "disconnected", connection, reason })
  }

  private notify(event: ConnectionChangeEvent): void {
    for (const subscriber of this.subscribers) {
      try {
        subscriber(event)
      } catch (error) {
        this.logger.warn({ err: error, eventType: event.type }, "Client connection subscriber failed")
      }
    }
  }
}

function getConnectionKey(input: ClientConnectionRef): string {
  return `${input.clientId}:${input.connectionId}`
}

File: packages/server-rust/src/clients/connection_manager.rs
use std::collections::HashMap;
use std::time::{Duration, Instant};

use crate::logger::Logger;

const STALE_CONNECTION_TIMEOUT: Duration = Duration::from_secs(45);

#[derive(Debug, Clone)]
pub struct ClientConnectionRef {
    pub client_id: String,
    pub connection_id: String,
}

#[derive(Debug, Clone)]
pub struct ClientConnectionRecord {
    pub client_id: String,
    pub connection_id: String,
    pub key: String,
    pub connected_at: Instant,
    pub last_seen_at: Instant,
}

#[derive(Debug, Clone)]
pub enum ConnectionChangeEvent {
    Connected { connection: ClientConnectionRecord },
    Disconnected { connection: ClientConnectionRecord, reason: String },
}

struct RegisteredConnection {
    record: ClientConnectionRecord,
    close: Option<Box<dyn FnOnce() + Send>>,
}

pub struct ClientConnectionManager {
    connections: HashMap<String, RegisteredConnection>,
    subscribers: Vec<Box<dyn Fn(&ConnectionChangeEvent) + Send + Sync>>,
}

impl ClientConnectionManager {
    pub fn new(_logger: Logger) -> Self {
        Self {
            connections: HashMap::new(),
            subscribers: Vec::new(),
        }
    }

    pub fn subscribe<F>(&mut self, listener: F)
    where
        F: Fn(&ConnectionChangeEvent) + Send + Sync + 'static,
    {
        self.subscribers.push(Box::new(listener));
    }

    pub fn register<F>(&mut self, input: ClientConnectionRef, close: F)
    where
        F: FnOnce() + Send + 'static,
    {
        let key = get_connection_key(&input);
        let now = Instant::now();

        if self.connections.contains_key(&key) {
            self.disconnect(&key, "replaced");
        }

        let record = ClientConnectionRecord {
            client_id: input.client_id.clone(),
            connection_id: input.connection_id.clone(),
            key: key.clone(),
            connected_at: now,
            last_seen_at: now,
        };

        self.connections.insert(key, RegisteredConnection {
            record: record.clone(),
            close: Some(Box::new(close)),
        });

        self.notify(&ConnectionChangeEvent::Connected { connection: record });
    }

    pub fn pong(&mut self, input: &ClientConnectionRef) -> bool {
        let key = get_connection_key(input);
        match self.connections.get_mut(&key) {
            Some(conn) => {
                conn.record.last_seen_at = Instant::now();
                true
            }
            None => false,
        }
    }

    pub fn is_connected(&self, input: &ClientConnectionRef) -> bool {
        let key = get_connection_key(input);
        self.connections.contains_key(&key)
    }

    pub fn sweep_stale(&mut self) {
        let cutoff = Instant::now() - STALE_CONNECTION_TIMEOUT;
        let stale_keys: Vec<String> = self
            .connections
            .iter()
            .filter(|(_, conn)| conn.record.last_seen_at < cutoff)
            .map(|(key, _)| key.clone())
            .collect();

        for key in stale_keys {
            self.disconnect(&key, "timeout");
        }
    }

    fn disconnect(&mut self, key: &str, reason: &str) {
        if let Some(mut conn) = self.connections.remove(key) {
            if let Some(close) = conn.close.take() {
                close();
            }
            self.notify(&ConnectionChangeEvent::Disconnected {
                connection: conn.record.clone(),
                reason: reason.to_string(),
            });
        }
    }

    fn notify(&self, event: &ConnectionChangeEvent) {
        for subscriber in &self.subscribers {
            subscriber(event);
        }
    }
}

fn get_connection_key(input: &ClientConnectionRef) -> String {
    format!("{}:{}", input.client_id, input.connection_id)
}
